### PR TITLE
Add keypair

### PR DESCRIPTION
**Pull request by szpotona**

### CODE REVIEW

File: plugin.yaml [hunk 1]
```diff
@@ -10,58 +10,215 @@ plugins:
     # URL to archive containing the plugin or name of directory containing
     # the plugin if it is included in the the blueprint directory under the
     # "plugins" directory.
-    source: https://github.com/cloudify-cosmo/cloudify-gcp-plugin/archive/master.zip
+    source: https://github.com/szpotona/cloudify-gcp-plugin/archive/add_keypair.zip
```
Comment (noakup) [inline plugin.yaml]: I guess this was for tests, but I think it should point to the master again

File: plugin.yaml [hunk 2]
```diff
@@ -10,58 +10,215 @@ plugins:
...
 
 node_types:
   cloudify.gcp.nodes.Instance:
     derived_from: cloudify.nodes.Compute
+    properties:
+      use_external_resource:
+        description: >
+          Indicate whether the resource exists or if Cloudify should create the resource.
```
Comment (noakup) [inline plugin.yaml]: It would be nice to explain what true/false mean (true - use existing, false - create)

File: plugin.yaml [hunk 3]
```diff
@@ -10,58 +10,215 @@ plugins:
...
     # the plugin if it is included in the the blueprint directory under the
     # "plugins" directory.
-    source: https://github.com/cloudify-cosmo/cloudify-gcp-plugin/archive/master.zip
+    source: https://github.com/szpotona/cloudify-gcp-plugin/archive/add_keypair.zip
 
 node_types:
   cloudify.gcp.nodes.Instance:
     derived_from: cloudify.nodes.Compute
+    properties:
+      use_external_resource:
```
Comment (noakup) [inline plugin.yaml]: I think the resource_id is missing here, but use_external_resource requires it... The same is also true for all other occurences of use_external_resource in this yaml

File: plugin.yaml [hunk 4]
```diff
@@ -10,58 +10,215 @@ plugins:
...
+        description: >
+          A dictionary of values to pass to authenticate with the Google Cloud Platform API.
+        default: {}
+      image_id:
+        description: >
+          The ID of the image in your GCP account.
+        type: string
+      instance_type:
+        description: >
+          The instance's size.
```
Comment (noakup) [inline plugin.yaml]: I think it's the flavor/whatever google is calling this property (size suggests it's a number...)

File: plugin.yaml [hunk 5]
```diff
@@ -10,58 +10,215 @@ plugins:
...
+        type: string
+        default: n1-standard-1
+      properties:
+        description: >
```
Comment (noakup) [inline plugin.yaml]: if those properties are only used by create, they should probably be defined only for the create operation. Also, try to use a different name (maybe "args"?) because it's nested in "properties" as well

File: plugin.yaml [hunk 6]
```diff
@@ -10,58 +10,215 @@ plugins:
...
-        default: ''
+      use_external_resource:
+        description: >
+          Indicate whether the resource exists or if Cloudify should create the resource.
+        type: boolean
+        default: false
+      gcp_config:
+        description: >
+          A dictionary of values to pass to authenticate with the Google Cloud Platform API.
+        default: {}
```
Comment (noakup) [inline plugin.yaml]: the default should probably be { get_property: [SELF, gcp_config] } like you did in cloudify.gcp.nodes.Instance, no?

File: plugin.yaml [hunk 7]
```diff
@@ -10,58 +10,215 @@ plugins:
...
+          Indicate whether the resource exists or if Cloudify should create the resource.
+        type: boolean
+        default: false
+      gcp_config:
+        description: >
+          A dictionary of values to pass to authenticate with the GCP API.
+        default: {}
+
 relationships:
   cloudify.gcp.relationships.connected_to:
```
Comment (noakup) [inline plugin.yaml]: The relationships types "cloudify.gcp.relationships.connected_to" and "cloudify.gcp.relationships.contained_in" are not used, so you can remove them

File: plugin.yaml [hunk 8]
```diff
@@ -69,3 +226,92 @@ relationships:
   cloudify.gcp.relationships.contained_in:
     derived_from: cloudify.relationships.contained_in
 
+  cloudify.gcp.relationships.agent_connected_to_security_group:
```
Comment (noakup) [inline plugin.yaml]: Seems like the only difference between these 2 relationships types is the tag (agent, manager). Since they have the same schema but differ in values, maybe you can just use a single relationship named "instance_connected_to_security_group" and set the tag as a dynamic input value.
(Also, what is the tag used for?)

File: plugin/gcp/instance.py [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        :param instance_name: name of the instance
+        :param image: image id in Google Cloud Platform
+        :param machine_type: machine type on GCP, default None
+        :param startup_script: shell script text to be run on instance startup,
+        default None
+        :param external_ip: boolean external ip indicator, default False
+        :param tags: tags for the instance, default []
+        """
+        super(Instance, self).__init__(config, logger)
+        self.project = config['project']
```
Comment (noakup) [inline plugin/gcp/instance.py]: I think this is redundant since it's done in the super's __init__ also

File: plugin/gcp/instance.py [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        self.zone = config['zone']
+        self.name = utils.get_gcp_resource_name(instance_name)
+        self.image = image
+        self.machine_type = machine_type \
+            if machine_type else self.STANDARD_MACHINE_TYPE
```
Comment (noakup) [inline plugin/gcp/instance.py]: if machine type is missing - we should probably fail.. (you defined a default value for that in the plugin.yaml, which is good because the user can see and change it there)

File: plugin/gcp/instance.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+    ACCESS_CONFIG = 'External NAT'
+    ACCESS_CONFIG_TYPE = 'ONE_TO_ONE_NAT'
+    NETWORK_INTERFACE = 'nic0'
+    STANDARD_MACHINE_TYPE = 'n1-standard-1'
+
+    def __init__(self,
+                 config,
+                 logger,
+                 instance_name,
+                 image=None,
```
Comment (noakup) [inline plugin/gcp/instance.py]: I think image and machine type must always have a value, so there's probably no need to set a default value here to None

File: plugin/gcp/instance.py [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        tag setting process and its status
+        """
+        # each tag should be RFC1035 compliant
+        self.logger.info('Set tags')
+        self.tags.extend(tags)
+        self.tags = list(set(self.tags))
+        fingerprint = self.get()["tags"]["fingerprint"]
```
Comment (noakup) [inline plugin/gcp/instance.py]: can we assume the fingerprint field is always included in tags?

File: plugin/gcp/instance.py [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+            project=self.project,
+            zone=self.zone,
+            instance=self.name,
+            body={"items": self.tags, "fingerprint": fingerprint}).execute()
+
+    def get(self):
+        """
+        Get GCP instance details.
+        """
+        self.logger.info("Get instance details")
```
Comment (noakup) [inline plugin/gcp/instance.py]: it would help if more details are printed to the log (all that we have - instance name, project and zone)

File: plugin/gcp/instance.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+    def remove_tags(self, tags):
+        """
+        Remove GCP instance tags.
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        tag removal process and its status
+        """
+        # each tag should be RFC1035 compliant
+        self.logger.info('Remove tags')
```
Comment (noakup) [inline plugin/gcp/instance.py]: More details logging would be helpful - which tags are removed, from which instance)

File: plugin/gcp/instance.py [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+    @blocking(True)
+    def delete(self):
+        """
+        Delete GCP instance.
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        deletion process and its status
+        """
+        self.logger.info('Delete instance')
```
Comment (noakup) [inline plugin/gcp/instance.py]: More details logging would be better - which instance is deleted? best to include all the details we have here  the instance, the project and the zone.

File: plugin/gcp/instance.py [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+        external IP setting process and its status
+        """
+        body = {"kind": "compute#accessConfig",
+                "name": self.ACCESS_CONFIG,
+                "type": self.ACCESS_CONFIG_TYPE}
+        return self.compute.instances().addAccessConfig(
+            project=self.project,
+            instance=self.name,
+            zone=self.zone,
+            networkInterface=self.NETWORK_INTERFACE,
```
Comment (noakup) [inline plugin/gcp/instance.py]: adding the external IP is always on nic0?  also will it always be an External NAT and 1-to-1? Are there other configuration we can support here?

File: plugin/gcp/instance.py [hunk 17]
```diff
@@ -0,0 +1,239 @@
...
+    def delete(self):
+        """
+        Delete GCP instance.
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        deletion process and its status
+        """
+        self.logger.info('Delete instance')
+        return self.compute.instances().delete(
```
Comment (noakup) [inline plugin/gcp/instance.py]: Here it might be required to list the existing instances first. The question is really if the error thrown by gcp will be clear enough. Also of course it should cause a non-recoverable error on Cloudify so we won't retry this again if the instance does not exist.

File: plugin/gcp/instance.py [hunk 18]
```diff
@@ -0,0 +1,239 @@
...
+                    'boot': True,
+                    'autoDelete': True,
+                    'initializeParams': {
+                        'sourceImage': self.image
+                    }
+                }
+            ],
+            'networkInterfaces': [
+                {'network': 'global/networks/{0}'.format(self.network)}],
+            'serviceAccounts': [
```
Comment (noakup) [inline plugin/gcp/instance.py]: What are the service accounts? I see we use a default email setting and default scopes urls, will these values always work? should they be configurable?

File: plugin/gcp/instance.py [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+                    {'key': 'bucket', 'value': self.project}]
+            }
+        }
+        if self.tags:
+            body['tags'] = {"items": self.tags}
+        if self.externalIP:
+            for item in body['networkInterfaces']:
+                if item['name'] == self.ACCESS_CONFIG:
+                    item['accessConfigs'] = [{'type': 'ONE_TO_ONE_NAT',
+                                              'name': 'External NAT'}]
```
Comment (noakup) [inline plugin/gcp/instance.py]: you created consts for those

File: plugin.yaml [hunk 20]
```diff
@@ -10,58 +10,215 @@ plugins:
...
+      cloudify.interfaces.lifecycle:
+        create: 
+          implementation: gcp_plugin.plugin.tasks.create_security_group
+          inputs: 
+            gcp_config:
+              default: { get_property: [SELF, gcp_config] }
+            rules:
+              default: { get_property: [SELF, rules] }
+        delete:          
+          implementation: gcp_plugin.plugin.tasks.delete_firewall_rule
```
Comment (noakup) [inline plugin.yaml]: delete_firewall_rule => delete_security_group

File: plugin.yaml (hunk 1 above)
Comment (szpotona) [inline plugin.yaml]: Yes, but just before merge. If I'm making changes I want to be sure it's working.

File: plugin.yaml (hunk 20 above)
Comment (szpotona) [inline plugin.yaml]: It's exactly the same function so I want to leave it as it is. If you think this is reasonable I'll make delete_security_group operation calling delete firewall rule.

File: plugin.yaml (hunk 7 above)
Comment (szpotona) [inline plugin.yaml]: ok

File: plugin.yaml (hunk 8 above)
Comment (szpotona) [inline plugin.yaml]: You're right. I was more concerned on getting manager blueprint working that the design itself.

Tag is used for security groups that are based on firewall rules. They differ from what you can find in Openstack and AWS so I had to create something that will behave the same way. I could explain you that on Skype.

File: plugin/gcp/instance.py (hunk 12 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: Yes, it's always there.

File: plugin/gcp/instance.py (hunk 16 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: It's the only way possible. Moreover GCP supports instance connected to only one network at the moment.

File: plugin/gcp/instance.py (hunk 19 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: Thanks.

File: plugin.yaml (hunk 3 above)
Comment (szpotona) [inline plugin.yaml]: It's not implemented. I had to add use external resource because boostrap manager process has some validation and it wasn't passing before. But it's not possible in some points: e.g. in KeyPair there is no sense to put it because Key doesn't have its name.

File: plugin/gcp/instance.py (hunk 11 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: There are many operations as listing the instances, removing them or setting tags to which I need to create Instance object. Look at tasks.py file, it's widely used there.

File: plugin/gcp/instance.py (hunk 9 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: I've missed that.Thanks

File: plugin/gcp/instance.py (hunk 10 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: OK

File: plugin/gcp/instance.py (hunk 17 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: The error thrown by GCP is really very detailed and I don't see need of checking whether the instance exists before. All tasks raise a non-recovery exception in case of GCPError.

File: plugin/gcp/service.py [hunk 21]
```diff
@@ -13,15 +13,32 @@
...
+import time
 
 import Crypto
 import httplib2
 from googleapiclient.discovery import build
 from oauth2client.client import SignedJwtAssertionCredentials
 
-from plugin.gcp import utils
+
+def blocking(default):
```
Comment (noakup) [inline plugin/gcp/service.py]: Maybe there's no need for a default value here, and using the blocking decorator is indicative enough? (I see you always use "@blocking(True)" )

File: plugin/gcp/service.py [hunk 22]
```diff
@@ -192,12 +107,12 @@ def wait_for_operation(self,
...
                     operation=operation).execute()
             else:
                 result = self.compute.zoneOperations().get(
-                    project=self.project['name'],
-                    zone=self.project['zone'],
+                    project=self.project,
+                    zone=self.zone,
                     operation=operation).execute()
             if result['status'] == 'DONE':
                 if 'error' in result:
```
Comment (noakup) [inline plugin/gcp/service.py]: please log more details on what failed (i.e. operation name, project, zone if relevant)

File: plugin/gcp/utils.py [hunk 23]
```diff
@@ -63,9 +53,18 @@ def get_gcp_resource_name(name):
...
         remain_len = MAX_GCP_INSTANCE_NAME - len(final_name)
         final_name = '{0}{1}'.format(final_name[:remain_len - ID_HASH_CONST],
                                      final_name[-ID_HASH_CONST:])
     # convert string to lowercase
     return final_name.lower()
+
+
+def get_firewall_rule_name(network, firewall):
+    """
+    Prefix firewall rule name with network name
```
Comment (noakup) [inline plugin/gcp/utils.py]: why did you remove the params docstring? it was good :smile:

File: plugin/integration_tests/blueprint/blueprint.yaml [hunk 24]
```diff
@@ -17,21 +22,19 @@ node_types:
                 start:
                     implementation: gcp_plugin.plugin.tasks.create_instance
                     inputs:
-                        config:
-                            default: {}
+                        gcp_config:
+                            default: { get_input: config }
+                        instance_type:
+                            default: n1-standard-1
```
Comment (noakup) [inline plugin/integration_tests/blueprint/blueprint.yaml]: since you added an input for this, use { get_input: instance_type } here and set the default value n1-standard-1 on the input itself

File: plugin/gcp/firewall.py [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+from plugin.gcp.service import GoogleCloudPlatform
+from plugin.gcp.service import blocking
+
+
+class FirewallRule(GoogleCloudPlatform):
+    def __init__(self,
```
Comment (noakup) [inline plugin/gcp/firewall.py]: I'd prefer init would only accept config and logger. other args (firewall and network or name and network) better be passed as args to specific methods (e.g. "create", "delete"), as required

File: plugin/gcp/instance.py (hunk 11 above)
Comment (noakup) [inline plugin/gcp/instance.py]: ok, I want to revise me comment :smile: 
I prefer only incuding the config and logger - which are required to initialize the super class - in the init method. Everything else - instance name, image, machine_type etc should be passed directly to the methods ("create", "list") when required. In each method declaration, mandatory args (such as "image" passed to "create"), should not have the default None.

File: plugin/gcp/instance.py (hunk 17 above)
Comment (noakup) [inline plugin/gcp/instance.py]: cool

File: plugin/gcp/instance.py (hunk 16 above)
Comment (noakup) [inline plugin/gcp/instance.py]: :+1:

File: plugin/gcp/keypair.py [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+                 private_key_path):
+        """
+        Create KeyPair object
+
+        :param config: gcp auth file
+        :param logger: logger object
+        :param user: name of the user to authenticate
+        :param private_key_path: path where private key is stored
+        """
+        super(KeyPair, self).__init__(config, logger)
```
Comment (noakup) [inline plugin/gcp/keypair.py]: here also I prefer passing the user and private_key_path args directly to the methods that require them and not to init

File: plugin/gcp/network.py [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+from plugin.gcp import utils
+from plugin.gcp.service import GoogleCloudPlatform
+from plugin.gcp.service import blocking
+
+
+class Network(GoogleCloudPlatform):
+    def __init__(self,
+                 config,
+                 logger,
+                 network):
```
Comment (noakup) [inline plugin/gcp/network.py]: Similar to the other classes - let's pass the network directly only to the methods that require it

File: plugin/gcp/keypair.py [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+            key = '{0}:{1}'.format(user, ssh_key)
+            if key not in item['value']:
+                item['value'] = '{0}\n{1}'.format(item['value'], key)
+        return self.compute.projects().setCommonInstanceMetadata(
+            project=self.project,
+            body=common_instance_metadata).execute()
+
+    @blocking(True)
+    def remove_project_ssh_key(self):
+        """
```
Comment (noakup) [inline plugin/gcp/keypair.py]: this method is not symmetric with add_project_ssh_key, which gets the public key as an argument.
anyways this will be settled by setting all the methods to get args and not use self.

File: plugin/gcp/instance.py [hunk 29]
```diff
@@ -0,0 +1,239 @@
...
+        """
+        Create GCP VM instance with given parameters.
+        Zone operation.
+
+        :return: REST response with operation responsible for the instance
+        creation process and its status
+        :raise: GCPError if there is any problem with startup script file:
+        e.g. the file is not under the given path or it has wrong permissions
+        """
+        self.logger.info('Create instance')
```
Comment (noakup) [inline plugin/gcp/instance.py]: how does the new instance create here get the public key to use? is it read from the common_instance_data? If so - what happens if there are multiple public keys there?

File: plugin/gcp/instance.py [hunk 30]
```diff
@@ -0,0 +1,239 @@
...
+
+    def to_dict(self):
+        body = {
+            'name': self.name,
+            'machineType': 'zones/{0}/machineTypes/{1}'.format(
+                self.zone,
+                self.machine_type),
+            'disks': [
+                {
+                    'boot': True,
```
Comment (noakup) [inline plugin/gcp/instance.py]: 1. I think these parameters (boot and autoDelete) should be configurable.
2. If a disk is not deleted after the instance is shut down, will another instance be able to use it? that could help us implement recovery.

File: plugin/gcp/instance.py [hunk 31]
```diff
@@ -0,0 +1,239 @@
...
+        return self.compute.instances().list(
+            project=self.project,
+            zone=self.zone).execute()
+
+    def wait_for_operation(self, operation, global_operation=False):
+        super(Instance, self).wait_for_operation(operation, global_operation)
+
+    def to_dict(self):
+        body = {
+            'name': self.name,
```
Comment (noakup) [inline plugin/gcp/instance.py]: can we also add a description here like you do for network? "Cloudify generated instance"

File: plugin/tasks.py [hunk 32]
```diff
@@ -35,103 +40,245 @@ def _decorator(*args, **kwargs):
...
+def save_private_key(gcp_config,
+                     private_key,
+                     user,
+                     private_key_path,
+                     **kwargs):
+    keypair = KeyPair(gcp_config,
+                      ctx.logger,
+                      user,
+                      private_key_path)
+    keypair.private_key = private_key
```
Comment (noakup) [inline plugin/tasks.py]: something missing here? I think this will have no effect

File: plugin.yaml [hunk 33]
```diff
@@ -10,62 +10,281 @@ plugins:
...
+              default: { get_property: [SOURCE, gcp_config] }
+            instance_name: 
+              default: { get_attribute: [SOURCE, gcp_name] }
 
+  cloudify.gcp.relationships.instance_connected_to_keypair:
+    derived_from: cloudify.relationships.connected_to
+    source_interfaces:
+      cloudify.interfaces.relationship_lifecycle:
+        establish:
+          implementation: gcp_plugin.plugin.tasks.save_private_key
```
Comment (noakup) [inline plugin.yaml]: I think you don't need to explicity define the establish step and make it call save_private_key, because the key pair is created and the private key is saved on the create_keypair operation. This probably means you can just declare in the manager blueprint that there's a relationship between the server node and the keypair node, and they will be created in the correct order.

File: plugin/tasks.py [hunk 34]
```diff
@@ -13,125 +13,264 @@
...
+def delete_keypair(gcp_config, user, private_key_path, **kwargs):
+    keypair = KeyPair(gcp_config,
+                      ctx.logger,
+                      user,
+                      private_key_path)
+    keypair.public_key = ctx.instance.runtime_properties['gcp_public_key']
+    keypair.remove_project_ssh_key()
+    ctx.instance.runtime_properties.pop('gcp_public_key')
+    ctx.instance.runtime_properties.pop('gcp_private_key')
+
```
Comment (noakup) [inline plugin/tasks.py]: I think the private key should be deleted here, symmetric to how it is saved at the end of create_keypair

File: plugin.yaml [hunk 35]
```diff
@@ -10,62 +10,281 @@ plugins:
...
+            gcp_config:
+              default: { get_property: [SOURCE, gcp_config] }
+            private_key: 
+              default: { get_attribute: [TARGET, gcp_private_key] }
+            user: 
+              default: { get_property: [TARGET, user] }
+            private_key_path: 
+              default: { get_property: [TARGET, private_key_path] }
+        unlink:
+          implementation: gcp_plugin.plugin.tasks.delete_private_key
```
Comment (noakup) [inline plugin.yaml]: same applies here (if you take into account my comment on tasks.py to call delete_private_key at the end of delete_keypair)

File: plugin/gcp/instance.py (hunk 29 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: All the public keys from commonInstanceMetadata are injected into new instances on startup.

File: plugin/gcp/instance.py (hunk 30 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: They will, I'm preparing the next branch 'add disk'. It will allow disk creation/deletion and later snapshots management.

File: plugin/gcp/instance.py (hunk 31 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: OK

File: plugin/tasks.py (hunk 32 above)
Comment (szpotona) [inline plugin/tasks.py]: I had to change some things to get boostrap working with this branch. I'll investigate it.

File: plugin/gcp/keypair.py (hunk 28 above)
Comment (szpotona) [inline plugin/gcp/keypair.py]: That's true, I've missed that.

File: plugin/integration_tests/blueprint/blueprint.yaml (hunk 24 above)
Comment (szpotona) [inline plugin/integration_tests/blueprint/blueprint.yaml]: OK, corrected.

File: plugin/gcp/utils.py (hunk 23 above)
Comment (szpotona) [inline plugin/gcp/utils.py]: Readded. Thanks.

File: plugin/gcp/service.py (hunk 21 above)
Comment (szpotona) [inline plugin/gcp/service.py]: You're right, moreover I've found that I only take this argument, not delete. To be more elegant I've replaced kwargs.get('blocking') with kwargs.pop('blocking', True).

File: plugin.yaml (hunk 20 above)
Comment (szpotona) [inline plugin.yaml]: I've rearraged segurity groups creation and instance tagging. Adding tags to runtime_properties in create_security_group made it reasonable to add delete_security_group that removes them.

File: plugin/gcp/instance.py (hunk 18 above)
Comment (szpotona) [inline plugin/gcp/instance.py]: They should. Please let me do that in different branch/PR.

File: plugin/gcp/service.py (hunk 21 above)
Comment (noakup) [inline plugin/gcp/service.py]: :clap:

File: plugin/gcp/instance.py (hunk 29 above)
Comment (noakup) [inline plugin/gcp/instance.py]: thanks
